add a extra option to round for floor/ceil/round with precision

Sadrak opened the pull request:
see #9

Sadrak committed:
add a extra option to round for floor/ceil/round with precision

see #9

billward commented:
I think we should add floor() and ceil() methods as well, which are just wrappers for round() with hard-coded third option.

Sadrak committed:
change roundoption check to exact int check for future options

see #9

Sadrak commented:
OK, i have changed the check so if desired we can use constants and add more options

billward commented:
Also I really think that we should support this in format_number etc. as well.

Sadrak committed:
- added absolute floor/ceil to roundoption
- floor(), abs_floor(), ceil() and abs_ceil() added
- roundoption integrated as fifth argument to format_number()
- see #9

Sadrak committed:
- Docfix
- see #9

Sadrak committed:
- Docfix
- see #9

billward commented:
This looks good, but I just want to add one more change ... instead of the hard coded numbers -2, -1, 0, 1, 2 let's add some new constants: ROUND_ABS_FLOOR, ROUND_FLOOR, ROUND_NORMAL, ROUND_CEIL, ROUND_ABS_CEIL or similar? And use those everywhere you currently have the hard coded numbers (including in the tests).

Sadrak commented:
OK, can you guide me how do you like constants?

use constant?
Readonly?

And should we export the constants? Or only for internal use?

2017-03-07 20:16 GMT+01:00 Bill Ward <notifications@github.com>:

> This looks good, but I just want to add one more change ... instead of the
> hard coded numbers -2, -1, 0, 1, 2 let's add some new constants:
> ROUND_ABS_FLOOR, ROUND_FLOOR, ROUND_NORMAL, ROUND_CEIL, ROUND_ABS_CEIL or
> similar? And use those everywhere you currently have the hard coded numbers
> (including in the tests).
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/billward/number-format-perl/pull/11#issuecomment-284826040>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AADacMe2yCDKR0uks7OWEN8Q3XV_-c02ks5rja0OgaJpZM4L2lWQ>
> .
>

billward commented:
Use constant would be fine - there is already one constant in the module
for MAX_INT.
They can be in @EXPORT_OK so the user has the choice.


On Tue, Mar 7, 2017 at 11:22 AM, Felix Ostmann <notifications@github.com>
wrote:

> OK, can you guide me how do you like constants?
>
> use constant?
> Readonly?
>
> And should we export the constants? Or only for internal use?
>
> 2017-03-07 20:16 GMT+01:00 Bill Ward <notifications@github.com>:
>
> > This looks good, but I just want to add one more change ... instead of
> the
> > hard coded numbers -2, -1, 0, 1, 2 let's add some new constants:
> > ROUND_ABS_FLOOR, ROUND_FLOOR, ROUND_NORMAL, ROUND_CEIL, ROUND_ABS_CEIL or
> > similar? And use those everywhere you currently have the hard coded
> numbers
> > (including in the tests).
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/billward/number-format-perl/pull/11#issu
> ecomment-284826040>,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/AADacMe2y
> CDKR0uks7OWEN8Q3XV_-c02ks5rja0OgaJpZM4L2lWQ>
> > .
> >
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/billward/number-format-perl/pull/11#issuecomment-284827600>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAU0ZCIFCdJTQhZRefnks8esAsIZHTy1ks5rja5cgaJpZM4L2lWQ>
> .
>



-- 
Check out my LEGO blog at brickpile.com <http://www.brickpile.com/>
Follow/friend me: Facebook <http://facebook.com/billward> • Flickr
<http://flickr.com/photos/billward/> • Twitter
<http://twitter.com/williamward> • LinkedIn
<http://www.linkedin.com/pub/william-ward/63/393/985/> • Goodreads
<https://www.goodreads.com/user/show/2554312-bill-ward>

Sadrak committed:
- use constants for round_option
- to export all constants use :constants or :all
- see #9

Sadrak commented:
Additional i make a :constants export (like :subs or :vars)

I hope the documentation changes from me are not so bad :-/

Also i hope i found all placed to explain the round_option.


Have a nice day!


2017-03-07 20:45 GMT+01:00 Bill Ward <notifications@github.com>:

> Use constant would be fine - there is already one constant in the module
> for MAX_INT.
> They can be in @EXPORT_OK so the user has the choice.
>
>
> On Tue, Mar 7, 2017 at 11:22 AM, Felix Ostmann <notifications@github.com>
> wrote:
>
> > OK, can you guide me how do you like constants?
> >
> > use constant?
> > Readonly?
> >
> > And should we export the constants? Or only for internal use?
> >
> > 2017-03-07 20:16 GMT+01:00 Bill Ward <notifications@github.com>:
> >
> > > This looks good, but I just want to add one more change ... instead of
> > the
> > > hard coded numbers -2, -1, 0, 1, 2 let's add some new constants:
> > > ROUND_ABS_FLOOR, ROUND_FLOOR, ROUND_NORMAL, ROUND_CEIL, ROUND_ABS_CEIL
> or
> > > similar? And use those everywhere you currently have the hard coded
> > numbers
> > > (including in the tests).
> > >
> > > —
> > > You are receiving this because you authored the thread.
> > > Reply to this email directly, view it on GitHub
> > > <https://github.com/billward/number-format-perl/pull/11#issu
> > ecomment-284826040>,
> > > or mute the thread
> > > <https://github.com/notifications/unsubscribe-auth/AADacMe2y
> > CDKR0uks7OWEN8Q3XV_-c02ks5rja0OgaJpZM4L2lWQ>
> > > .
> > >
> >
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/billward/number-format-perl/pull/11#
> issuecomment-284827600>,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/
> AAU0ZCIFCdJTQhZRefnks8esAsIZHTy1ks5rja5cgaJpZM4L2lWQ>
> > .
> >
>
>
>
> --
> Check out my LEGO blog at brickpile.com <http://www.brickpile.com/>
> Follow/friend me: Facebook <http://facebook.com/billward> • Flickr
> <http://flickr.com/photos/billward/> • Twitter
> <http://twitter.com/williamward> • LinkedIn
> <http://www.linkedin.com/pub/william-ward/63/393/985/> • Goodreads
> <https://www.goodreads.com/user/show/2554312-bill-ward>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/billward/number-format-perl/pull/11#issuecomment-284837208>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AADacBdJ1jd3x3dx-w2BccxJo0JbvgWbks5rjbPwgaJpZM4L2lWQ>
> .
>

Sadrak committed:
- tests for format_number also use round_option constants
- see #9

billward commented on Format.pm:
```diff
@@ -242,6 +256,7 @@ our $N_SIGN_POSN        = 1;    # sign rules for negative: 0-4
 our $DECIMAL_DIGITS     = 2;
 our $DECIMAL_FILL       = 0;
 our $NEG_FORMAT         = '-x';
+our $ROUND_OPTION       = 0;
```
should be ROUND_NORMAL - and move constants to above this point

Sadrak replied:
done

billward commented on Format.pm:
```diff
@@ -127,6 +129,11 @@ positive representation of the number being passed to that function.
+C<ROUND_OPTION> is only used by C<round()> or C<format_number()>.
+-1 or ROUND_FLOOR means floor, -2 or ROUND_ABS_FLOOR means absolute
+floor, 1 or ROUND_CEIL means ceil, 2 or ROUND_ABS_CEIL means absolute
+ceil and a 0 or ROUND_NORMAL stand for normal rounding.
```
change 'stand for' to 'means'. Also add "see the methods abs_floor(), floor(), abs_ceil(), and ceil() for more information"

billward commented on Format.pm:
```diff
@@ -482,18 +506,25 @@ sub new
+value 0). Examples:
+
+  round(3.14159)                  yields    3.14
+  round(3.14159, undef, 1)        yields    3.15
```
change 1 & -1 to ROUND_xx constants
Document that undef means to use the default precision (DECIMAL_DIGITS setting).

Sadrak replied:
added a passage for undef as argument

billward commented on Format.pm:
```diff
@@ -505,24 +536,50 @@ variables, use C<format_number()> instead.
     {
         my $rounded = $number->copy();
-        $rounded->precision(-$precision);
+
```
Let's add this:
my $sign = $rounded->is_neg() ? -1 : 1;
and then not call is_neg() repeatedly

Sadrak replied:
OK, for ROUND_ABS_FLOOR and ROUND_ABS_CEIL it is possible to call is_neg() twice. But to set $sign correctly we need to call is_neg() and is_pos() to get $sign = -1|0|1 (like $sign at other parts in the code).


```
            my $sign = $rounded->is_neg() ? -1
                     : $rounded->is_pos() ?  1
                     :                       0;
```

Additional i think the options ROUND_ABS_FLOOR & ROUND_ABS_CEIL are rarely used, so better call is_neg() twice when this options are set then call is_neg() every time without the need (for ROUND_FLOOR, ROUND_NORMAL and ROUND_CEIL).

billward replied:
ok I just dislike calling the same method multiple times even if it is a pretty efficient call.

Sadrak replied:
Me to, but calling a method without using the result also produce strong feelings :)

And i think maintainable code is more important than writing 10 more lines of code to do the correct thing to avoid a possible second function call (in a really rare case).

billward commented on Format.pm:
```diff
@@ -536,14 +593,124 @@ sub round
-    $result = -$result if $sign < 0;
+    if (
+        $roundoption == ROUND_FLOOR
+        or ($roundoption == ROUND_ABS_FLOOR and not $sign < 0)
```
change 'not $sign < 0' to '$sign >= 0'

Sadrak replied:
done, but now it is not consistent with the behavior from the Math::BigFloat part (`not $rounded->is_neg()`

billward commented on Format.pm:
```diff
@@ -553,21 +720,24 @@ specifier; trailing zeroes will only appear in the output if
+  format_number(1234567.8, 2, 1)                       yields  '1,234,567.80'
+  format_number(1.23456789, 6)                         yields  '1.234568'
+  format_number("0.000020000E+00", 7);                 yields  '2e-05'
+  format_number(1.3579, 2, undef, undef, ROUND_FLOOR)  yields  '1.35'
```
explain that passing undef causes the default value to be used for those items

Sadrak replied:
added a passage for undef as argument

billward reviewed (changes requested):
Overall this is very good. I have a couple of minor suggestions that I made as comments in the code. I'll merge it once those are addressed.

Sadrak committed:
- use `$sign >= 0` instead of `not $sign < 0`
- Documentation changes
 - explain undef as argument to round() and format_number()
- see #9
